internal/controller/pod_webhook.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+//+kubebuilder:webhook:path=/mutate-v1-pod,mutating=true,failurePolicy=fail,sideEffects=None,groups="",resources=pods,verbs=create,versions=v1,name=instaslice.codeflare.dev,admissionReviewVersions=v1
+
+type PodAnnotator struct {
+	Client  client.Client
+	Decoder *admission.Decoder
+}
+
+func (a *PodAnnotator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	// Decode the incoming pod object
+	pod := &v1.Pod{}
+	err := a.Decoder.Decode(req, pod)
+	if err != nil {
+		return admission.Errored(400, fmt.Errorf("could not decode pod: %v", err))
+	}
+
+	// Get the pod name from the request object
+	podName := pod.Name
+	if podName == "" {
+		return admission.Errored(400, fmt.Errorf("pod name should not be empty"))
+	}
+
+	if !hasMIGResource(pod) {
+		return admission.Allowed("No nvidia.com/mig-* resource found, skipping mutation.")
+	}
+
+	// Add finalizer
+	finalizerName := "org.instaslice/accelarator"
+	if !containsString(pod.Finalizers, finalizerName) {
+		pod.Finalizers = append(pod.Finalizers, finalizerName)
+	}
+
+	// Add scheduling
+	schedulingGateName := "org.instaslice/accelarator"
+	found := false
+	for _, gate := range pod.Spec.SchedulingGates {
+		if gate.Name == schedulingGateName {
+			found = true
+			break
+		}
+	}
+	if !found {
+		pod.Spec.SchedulingGates = append(pod.Spec.SchedulingGates, v1.PodSchedulingGate{Name: schedulingGateName})
+	}
+
+	// Generate an extended resource name based on the pod name
+	uuidStr := uuid.New().String()
+	extendedResourceName := fmt.Sprintf("org.instaslice/%s-%s", podName, uuidStr)
```

Comment:
This is used in multiple places, I think it should be a constant (esp. in case you'll want to change it later).

Reply:
agree this is all over the code, let me create an issue for it and lets us come back to it later.

Reply:
opened #21